pallets/compliance-manager/src/lib.rs

```diff
@@ -780,33 +779,39 @@ impl<T: Trait> ComplianceManagerTrait<T::Balance> for Module<T> {
         from_did_opt: Option<IdentityId>,
         to_did_opt: Option<IdentityId>,
         _value: T::Balance,
+        treasury_did: Option<IdentityId>,
     ) -> StdResult<u8, &'static str> {
-        // Transfer is valid if ALL receiver AND sender rules of ANY asset rule are valid.
+        let (treasury_sender, treasury_receiver) = if treasury_did.is_some() {
+            (treasury_did == from_did_opt, treasury_did == to_did_opt)
+        } else {
+            (false, false)
+        };
+        // Transfer is valid if ALL receiver AND sender rules of ANY asset rule are valid. An
+        // exception is made for the cases when either the sender or the receiver is the treasury
+        // DID.
         let asset_rules = Self::asset_rules(ticker);
-        if asset_rules.is_paused {
+        if asset_rules.is_paused
+            || (asset_rules.rules.len() == 0 && (treasury_sender || treasury_receiver))
```

Comment:
This could also be done by moving:
`let mut rule_satisfied = false;` outside of the `for` loop and initialising it to `true` rather than `false`.

Reply:
@vkomenda  Yeah still need to verify the asset rule for its (treasury DID) companion sender/receiver.

Reply:
> I'm not 100% sure about this logic. IIUC it means that if the sender or receiver is the treasuryDID AND there are no asset rules, the transfer will succeed.
> 
> This changes the previous behaviour that no asset rules means a failure, not success. Even if the sender / receiver is the treasury DID, the other side (i.e. receiver / sender) needs to be checked for compliance, so my extension this would mean a failure if there are no asset rules.

@adamdossa In fact the previous behaviour was to pass any transaction given this always true rule:
```rust
ComplianceManager::add_active_rule(
            owner_signed.clone(),
            ticker,
            vec![],
            vec![]
        )
```
where the lists of sender and receiver rules are empty. I preserved the action of this rule so that the unit tests that use this rule would still pass.

Correct me if I'm wrong, I understood the new requirements that we have to make the check pass when there are no rules and either the sender or the receiver is the treasury DID. This is what the amended condition achieves.

Reply:
> @vkomenda Yeah still need to verify the asset rule for its (treasury DID) companion sender/receiver.

@satyamakgec I don't understand your comment. When `asset_rules.rules.len() == 0` there are no sender or receiver rules. This new clause is added to support the always true "rule" (we need to fix the terminology which is ambiguous at the moment).

Reply:
As I understand it, the case where we have:
```
ComplianceManager::add_active_rule(
            owner_signed.clone(),
            ticker,
            vec![],
            vec![]
        )
```
would mean that `asset_rules.rules.len() == 1` and so not relevant for this clause (which is just checking when `asset_rules.rules.len() == 0`.

If there are no asset rules (`asset_rules.rules.len() == 0`) then we should only allow the transfer is:
`sender == None AND receiver == TreasuryDID` OR `sender == TreasuryDID AND receiver == None`
this covers the issuance / redemption cases and is consistent with the previous behaviour plus this change.

Reply:
When we have `asset_rules.rules.len() == 1` with a single empty rule (as per this test case) then the compliance_manager will always pass the transaction as `are_all_rules_satisfied` will always return true.

Reply:
I agree the terminology is very confusing here with regards to rules, as they appear at different levels but are always called rules.

Reply:
> If there are no asset rules (`asset_rules.rules.len() == 0`) then we should only allow the transfer if:
> `sender == None AND receiver == TreasuryDID` OR `sender == TreasuryDID AND receiver == None`
> this covers the issuance / redemption cases and is consistent with the previous behaviour plus this change.

Added this condition.